assigner/backends/base.py

```diff
@@ -0,0 +1,154 @@
+
+    @property
+    def namespace_id(self) -> str:
+        raise NotImplementedError
+
+    @property
+    def id(self) -> str:
+        raise NotImplementedError
+
+    @property
+    def info(self) -> Optional[str]:
+        raise NotImplementedError
+
+    @property
+    def repo(self) -> Optional[git.Repo]:
+        raise NotImplementedError
+
+    @property
+    def ssh_url(self) -> str:
+        raise NotImplementedError
+
+    def already_exists(self) -> bool:
+        raise NotImplementedError
+
+    def get_head(self, branch: str):
+        raise NotImplementedError
+
+    def checkout(self, branch: str):
+        raise NotImplementedError
+
+    def pull(self, branch: str) -> None:
+        raise NotImplementedError
+
+    def clone_to(self, dir_name: str, branch: Optional[str]) -> None:
+        raise NotImplementedError
+
+    def add_local_copy(self, dir_name: str) -> None:
+        raise NotImplementedError
+
+    def delete(self) -> None:
+        raise NotImplementedError
+
+    @classmethod
+    def get_user_id(cls, username: str, url_base: str, token: str) -> str:
+        raise NotImplementedError
+
+    def list_members(self) -> str:
+        raise NotImplementedError
+
+    def get_member(self, user_id: str) -> str:
+        raise NotImplementedError
+
+    def add_member(self, user_id: str, level: str) -> str:
+        raise NotImplementedError
+
+    def edit_member(self, user_id: str, level: str) -> str:
+        raise NotImplementedError
+
+    def delete_member(self, user_id: str) -> str:
+        raise NotImplementedError
+
+    def list_commits(self, ref_name: str = "master") -> str:
+        raise NotImplementedError
+
+    def list_pushes(self) -> str:
+        raise NotImplementedError
+
+    def get_last_HEAD_commit(self, ref: str = "master") -> str:
+        raise NotImplementedError
+
+    def list_branches(self) -> str:
+        raise NotImplementedError
+
+    def get_branch(self, branch: str) -> str:
+        raise NotImplementedError
+
+    def archive(self) -> str:
+        raise NotImplementedError
+
+    def unarchive(self) -> str:
+        raise NotImplementedError
+
+    def protect(self, branch: str = "master", developer_push: bool = True, developer_merge: bool = True) -> str:
+        raise NotImplementedError
+
+    def unprotect(self, branch: str = "master") -> str:
+        raise NotImplementedError
+
+
+class StudentRepoBase(RepoBase):
+    """Repository for a student's solution to a homework assignment"""
+
+    @classmethod
+    def new(cls, base_repo: str, semester: str, section: str, username: str, token: str):
+        raise NotImplementedError
+
+    @classmethod
+    def build_name(cls, semester: str, section: str, assignment: str, user: str) -> str:
+        raise NotImplementedError
+
+    def push(self, base_repo, branch: str = "master") -> None:
+        """Push base_repo code to this repo"""
+        raise NotImplementedError
+
+
+class TemplateRepoBase(RepoBase):
```

Comment:
Open for better names. `BaseRepoBase` didn't sound good. `HomeworkRepoBase` didn't clear things up for me either.

Reply:
I like `TemplateRepo`! We should make a note to update the wording in the documentation when we change this.

Reply:
Changed. I updated documentation where I could find it.